### ISSUE TITLE
Feature request: configure SSL settings in CHANNEL_LAYERS

### ISSUE THREAD

**Issue by jray-afk** [django/channels_redis#235]
Hi, I posted about my adventure trying to configure SSL for Redis in the suggested Google Groups page which has all the requested information prepopulated in the body of this issue form, haven't gotten a response yet: https://groups.google.com/g/django-users/c/S094vik_gW0

After digging into the channels_redis code in this repo I realized that I probably need to submit a new feature request here in order to configure SSL settings in CHANNEL_LAYERS. (Unless, do you know of another way to configure SSL using channels_redis?)

Thank you!

**Comment by carltongibson** [comment]
So... we'd need to by able to provide an `SSLContext` when creating connections. 

See [`aioredis.create_connection()`](https://aioredis.readthedocs.io/en/v1.3.0/api_reference.html#aioredis.create_connection). 

Happy to accept input here.

**Comment by jray-afk** [comment]
@carltongibson I appreciate the response. Yes that makes sense.

For now I will set up symmetric_encryption_keys to add some level of protection, but ultimately I would prefer setting up SSL if possible.

**Comment by carltongibson** [comment]
Super! If you get started and need some input, let me know. Thanks! 👍

**Comment by skadz** [comment]
I was just running in to this issue myself today, so happy to help test!

**Comment by jray-afk** [comment]
@skadz glad to hear I'm not the only one! Haven't had a chance to fork off of this project and try to implement the changes, but I may have to go rogue and go off of the version I built my project on, 2.3.2

**Comment by skadz** [comment]
> @skadz glad to hear I'm not the only one! Haven't had a chance to fork off of this project and try to implement the changes, but I may have to go rogue and go off of the version I built my project on, 2.3.2

@jray-afk Does this mean you actually have made the changes and got it working and just need to do the fork/submit PR or that you hacked some sort of other "fix" in?  I'd like to get this implemented ASAP and if you have something that is working, I'd love to help you get it in to the main code, but also can help test anything you've done so far (or would like to know about how you hacked it in ;)

**Comment by nmacianx** [comment]
@jray-afk @skadz any updates here? I'm facing the very same issue and help would be much appreciated!!

**Comment by carltongibson** [comment]
Just awaiting someone to pick it up and submit a pull request I think

**Comment by dablak** [comment]
When configuring the hosts in the Django settings you can use a dictionary and pass the SSL context option amongst [other parameters](https://github.com/aio-libs/aioredis/blob/db88d9610caab0a7aa290abfef684b6babc2d109/aioredis/connection.py#L53) :

``` python
# Mute the Heroku Redis error:
# "[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain"
# by disabling hostname check.

ssl_context = ssl.SSLContext()
ssl_context.check_hostname = False

heroku_redis_ssl_host = {
    'address': 'rediss://:password@127.0.0.1:6379/0'  # The 'rediss' schema denotes a SSL connection.
    'ssl': ssl_context
}

CHANNEL_LAYERS = {
    'default': {
        'BACKEND': 'channels_redis.core.RedisChannelLayer',
        'CONFIG': {
            'hosts': (heroku_redis_ssl_host,)
        }
    },
}
```

**Comment by carltongibson** [comment]
@dablak Thanks for the comment — should get people going — but if we can `ssl_context.check_hostname = False`, surely we can configure the `SSLContext` correctly too? 🤔

Is this just documentation?

**Comment by nasir733** [comment]
@Amplitude88  the above solution stopped working for me as well were you able to solve this issue?

**Comment by Amplitude88** [comment]
> @Amplitude88 the above solution stopped working for me as well were you able to solve this issue?

Hey mate, yes i was in able to solve it. I've downgraded to 3.4.1
In Version 4 the problem is not solvable with the provided solution above.

Best regards,
Nils

**Comment by nasir733** [comment]
@Amplitude88  Yeah it also stopped working for me after I upgraded to channels v4 but it was working for a few days I think or i might have not noticed it since I don't have unit testing for my project

**Comment by nasir733** [comment]
@carltongibson  @ezeakeal @andrewgodwin  @adamchainz  do you know why it might have stopped working? and how can we fix it ?

**Comment by ezeakeal** [comment]
I have no info to add, just posted a working config after days of tweaks
haha

**Comment by carltongibson** [comment]
@nasir733 please don't @mention whole lists of people to try and draw attention to your issue. It's just spam. 

You need to look at how connection params are passed down to redis-py, and from there your particular redis provider (be it local or hosted or ...) 

Once you've worked that out you may be in a position to make a documentation PR to resolve this issue.

**Comment by nasir733** [comment]
> @nasir733 please don't @mention whole lists of people to try and draw attention to your issue. It's just spam.
> 
> You need to look at how connection params are passed down to redis-py, and from there your particular redis provider (be it local or hosted or ...)
> 
> Once you've worked that out you may be in a position to make a documentation PR to resolve this issue.

thanks for the immediate response I will try to fix the problem and make a PR. I might need some help since it will be my first PR to a major opensource project 
sorry 😔 for tagging so many people

**Comment by aerickson1337** [comment]
I dug into this a little after having similar issues to Amplitude88 and having to downgrade to 3.4.x

This seems like it might actually be a bug in the asyncio.connection Connection class? in either the `asyncio.connection` `SSLConnection` class or the `asyncio.connection` `RedisSSLContext` class. There's a slot defined in the asyncio `Connection` class for `ssl_context` but no arg for it in the init() so it's always ignored / set to `None`. Similar story in `RedisSSLContext` and `SSLConnection` where there are other SSL args but no arg for ssl_context so it ends up being `None` or ignored always as well. 

Probably missing something somewhere, but initial thoughts are it might be a pretty easy fix upstream in asyncio since they just need to allow for passing in ssl_context in either `RedisSSLContext`, `SSLConnection`, or `Connection`? Currently it seems like if you pass in ssl_context all of the asyncio classes are configured to just drop it on the ground and set it to `None` which is whats causing people problems. 

It seems strange to me that if you specifically pass in the individual `SSLContext` parameters e.g. `ssl_keyfile`, `ssl_certfile`, etc those are configured correctly and get handled by `SSLConnection` and `RedisSSLContext`, but the `ssl_context` specifically even though referenced and used in those classes is just dropped on the ground or built from scratch instead of allowing a passed in `ssl_context`. 

If anyone can sanity check this or has ideas why asyncio might have done it that way would be happy to hear anything on it. If anyone can confirm they think this is an upstream bug in asyncio I can try to make some time to put together a repro for an upstream asyncio issue post.

TLDR: Seems like possible asyncio bug?

**Comment by marcinowski** [comment]
I've submitted a PR for a similar issue last week which I think might solve the problem with current versions https://github.com/django/channels_redis/pull/337

**Comment by aerickson1337** [comment]
Hey marcinowski, that PR looks good and is what I did in my first attempt at getting it working. I think it should solve the problem but it would break the previous channels_redis SSL API from 3.4.x that let you pass in an SSLContext directly. From more digging around in redis-py and their implementation it seems like they don't support passing an SSLContext directly. I did find a work around for people who need to maintain that functionality of passing the SSLContext directly instead of as params, but it also changes the existing channels_redis channel layer API compared to older versions which I wasn't a fan of either.

monkeypatch the SSL classes used in redis-py to allow ssl_context 
```python
from redis.asyncio.connection import Connection, RedisSSLContext
from typing import Optional

class CustomSSLConnection(Connection):
    def __init__(
        self,
        ssl_context: Optional[str] = None,
        **kwargs,
    ):
        super().__init__(**kwargs)
        self.ssl_context = RedisSSLContext(ssl_context)

class RedisSSLContext:
    __slots__ = (
        "context",
    )

    def __init__(
        self,
        ssl_context,
    ):
        self.context = ssl_context

    def get(self):
        return self.context
```

using in channel declaration/connection
```python
import ssl
from NameOfOurMonkeyPatchFile import CustomSSLConnection
context = ssl.create_default_context(purpose=ssl.Purpose.CLIENT_AUTH, cafile='some.crt')
context.load_cert_chain(certfile='some.crt', keyfile='some.key')
channel_settings = {
    'CHANNEL_LAYERS': {
        'default': {
            'BACKEND': 'channels_redis.core.RedisChannelLayer',
            'CONFIG': {
                "hosts": [
                    {
                        'host': f"{os.environ['REDIS']}",
                        'port': 6379,
                        'connection_class': CustomSSLConnection,
                        'ssl_context': context,
                    }
                ],
            },
        },
    }
}
```

I think the solution marcinowski provided is probably the best solution moving forward if redis-py doesn't want to support passing in an SSLContext directly and instead doing it with configuration options. 

Either way seems like 4.0 ends up changing how the redis SSL connection API in the channel_layer declaration works. Not sure what the best path forward is but seems like there needs to be: 3.4.x SSL connection docs, 4.x SSL connection docs, and 3.4.x->4.x SSL changes docs. 

I could probably take a stab at updating the docs with that but would probably need to wait and see what the verdict is on macrinowski's PR -> #337 before it would make sense to write those up.

**Comment by paulina7m** [comment]
I've spent many days trying to combine all suggestions I found on resolving this issue on Heroku with Channels 4.0.0.
What worked for me, is slightly modified version of @aerickson1337 solution.

`from redis.asyncio.connection import Connection, RedisSSLContext
from typing import Optional
import ssl
from urllib.parse import urlparse

class CustomSSLConnection(Connection):
    def __init__(
        self,
        ssl_context: Optional[str] = None,
        **kwargs,
    ):
        super().__init__(**kwargs)
        self.ssl_context = RedisSSLContext(ssl_context)

class RedisSSLContext:
    __slots__ = (
        "context",
    )

    def __init__(
        self,
        ssl_context,
    ):
        self.context = ssl_context

    def get(self):
        return self.context


url = urlparse(os.environ.get("REDIS_URL"))

ssl_context = ssl.SSLContext()
ssl_context.check_hostname = False

CHANNEL_LAYERS = {
    'default': {
        'BACKEND': 'channels_redis.core.RedisChannelLayer',
        'CONFIG': {
            'hosts': [
                    {
                        'host': url.hostname,
                        'port': url.port,
                        'username': url.username,
                        'password': url.password,
                        'connection_class': CustomSSLConnection,
                        'ssl_context': ssl_context,
                    }
                ],
        }
    },
}
`

ah, it feels so good to finally see Django Channels working properly with Heroku Data for Redis.
Thank you everyone on this thread.

**Comment by sevdog** [comment]
Just to add an other info here: redis-py has not yet documented "how to provide kwargs to perform an SSL/TLS connection" (see https://github.com/redis/redis-py/issues/780).

There are already currently an [`SSLConnection`](https://github.com/redis/redis-py/blob/d95d8a24ed2af3eae80b7b0f14cbccc9dbe86e96/redis/asyncio/connection.py#L939) and a [`RedisSSLContext`](https://github.com/redis/redis-py/blob/d95d8a24ed2af3eae80b7b0f14cbccc9dbe86e96/redis/asyncio/connection.py#L985) class in redis-py (which are used by the `from_url` method). So there is no need to create/use custom classes.

**Comment by carltongibson** [comment]
This should be resolved by https://github.com/django/channels_redis/pull/370.

**Comment by carltongibson** [comment]
Copying @vsobotka's https://github.com/django/channels_redis/issues/373#issue-1953892079 for future searchers. 

> Earlier this year we upgraded redis-py for our app on Heroku from 3.9 to 4.5.4. We immediately ran into https://github.com/django/channels_redis/issues/235 and had to place in a workaround, with the ssl context. Everything went fine, until now, when we upgraded redis-py again, to 5.0.1. We immediately got "Connection reset by peer" error, preventing a stable connection. I was not able to find anything that would help. What helped us was to remove the workaround from https://github.com/django/channels_redis/issues/235 and use the newly supported option ssl_cert_reqs in layer config. So far this looks like it works perfectly well.

**Comment by alex-atkins** [comment]
@carltongibson  Could you give an example of your CHANNEL_LAYERS?  I'm having the same "Connection reset by peer" issue, but when I added ssl_cert_reqs, I get this error from channels: 
`TypeError: AbstractConnection.__init__() got an unexpected keyword argument 'ssl_cert_reqs`'

I'm using:
Django==4.2
channels==4.1
channels-redis==4.2

**Comment by bigfootjon** [comment]
@alex-atkins can you post your channel layers and confirm that it matches the way the [README says to configure this](https://github.com/django/channels_redis?tab=readme-ov-file#hosts)?

**Comment by alex-atkins** [comment]
@bigfootjon It turns out the exact configuration in the README does work, I had just ignored it because it said it was self-signed.  I was trying to utilize an SSL cert chain (not self-signed).  Do you know the correct way to do this?

**Comment by bigfootjon** [comment]
@alex-atkins can you post the full stack trace?
